apis/r/R/utils-matrixZeroBasedView.R

```diff
@@ -1,109 +1,94 @@
-#   vector/matrix is ONE-based.
-# - as.one.based(M0) returns M1.
-# - The only other supported operations are: dim(M0), nrow(M0), ncol(M0).
-
-#' Zero-based matrix shim
-#'
-#' @param one_based_matrix matrix or Matrix::sparseMatrix
-#'
-#' @return Shim providing elemental access to the matrix using zero-based indexes.
-#' @export
-matrixZeroBasedView <- function(one_based_matrix) {
-  if (!inherits(one_based_matrix, "matrix") && !inherits(one_based_matrix, "sparseMatrix")) {
-    stop("Matrix object must inherit class matrix or Matrix::sparseMatrix.")
-  }
-  structure(list(one_based_matrix = one_based_matrix), class = "matrixZeroBasedView")
-}
-
-#' Zero-based matrix element access
-#'
-#' @param x The zero-based matrix view.
-#' @param i Row index (zero-based).
-#' @param j Column index (zero-based).
-#' @param drop Coerce result to lowest possible dimension.
-#'
-#' @return The specified matrix elements. Vector- or matrix-valued results are returned
-#'         as conventional one-based R objects.
-#' @export
-`[.matrixZeroBasedView` <- function(x, i, j, drop = TRUE) {
-  if (missing(i) && missing(j)) {
-    x$one_based_matrix[, , drop = drop]
-  } else if (missing(i)) {
-    x$one_based_matrix[, j + 1, drop = drop]
-  } else if (missing(j)) {
-    x$one_based_matrix[i + 1, , drop = drop]
-  } else {
-    x$one_based_matrix[i + 1, j + 1, drop = drop]
-  }
-}
-
-#' dim
-#'
-#' @param x The zero-based matrix view.
-#'
-#' @return The dimensions of the matrix.
-#' @export
-dim.matrixZeroBasedView <- function(x) {
-  dim(x$one_based_matrix)
-}
-
-#' nrow
-#'
-#' @param x The zero-based matrix view.
+#' matrixZeroBasedView is a wrapper shim for a matrix or Matrix::sparseMatrix providing
 #'
-#' @return Matrix row count.
+#' @description
+#' `matrixZeroBasedView` is a class that allows elemental 
+#'  matrix access using zero-based indeces.
 #' @export
-nrow.matrixZeroBasedView <- function(x) {
-  nrow(x$one_based_matrix)
-}
 
-#' ncol
-#'
-#' @param x The zero-based matrix view.
-#'
-#' @return Matrix column count.
-#' @export
-ncol.matrixZeroBasedView <- function(x) {
-  ncol(x$one_based_matrix)
-}
-
-#' Get one-based object
-#'
-#' @param x The object.
-#'
-#' @return A one-based version/view of the object.
-#' @export
-as.one.based <- function(x) {
-  UseMethod("as.one.based")
-}
+matrixZeroBasedView <- R6::R6Class(
+  classname = "matrixZeroBasedView",
+  public = list(
+  
+    #' @description Initialize (lifecycle: experimental)
+    #' @param x \link{matrix} or Matrix::\link[Matrix]{sparseMatrix} or Matrix::\link[Matrix]{denseMatrix}
+    initialize = function(x) {
+      if (!inherits(x, "matrix") && !inherits(x, "sparseMatrix") && !inherits(one_based_matrix, "denseMatrix")) {
+        stop("Matrix object must inherit class matrix or Matrix::sparseMatrix or Matrix:denseMatrix")
+      }
+      if (length(dim(x)) != 2) {
+        stop("Only two-dimensional matrices are supported")
+      }
```

Comment:
Just as an FYI we have in many other places switched to using the more compact `stopifnot` where the named argument _is_ the error message.  So this would be (untested, and I sometimes get the boolean inversion wrong on the first try)

```r
stopifnot("Matrix object must inherit class matrix or Matrix::sparseMatrix or Matrix:denseMatrix" =
      inherits(x, "matrix") || inherits(x, "sparseMatrix") || inherits(one_based_matrix, "denseMatrix",
      "Only two-dimensional matrices are supported" = length(dim(x)) == 2)
```